pintc/src/predicate/optimize/dead_code_elimination.rs

```diff
@@ -53,12 +53,92 @@ pub(crate) fn dead_state_elimination(contract: &mut Contract) {
 
 /// Remove all trivial Constraints in their respective predicates.
 ///
+/// Simplify any constraint conditions that consist of constant binary operations. Ex:
+/// true || <expr> is true
+/// true && <expr> is <expr>
+/// false || <expr> is <expr>
+/// false && <expr> is <false>
+///
 /// If any constraint evaluates to false, all constraints are removed and replaced with a single
 /// instance of `constraint false`
 pub(crate) fn dead_constraint_elimination(handler: &Handler, contract: &mut Contract) {
     let evaluator = Evaluator::new(contract);
 
     for pred_key in contract.preds.keys().collect::<Vec<_>>() {
```

Comment:
I added the optimization to `dead_constraint_elimination` for two reasons.

1. Optimization passes are supposed to be independent of other passes. By simplifying in this pass, I didn't have to repeat any code for the cases of `true || <expr> is true` and `false && <expr> is <false>`. Otherwise I would have to make sure the new simplification took place before this pass or repeat the pass to handle `constraint false` and `constraint true`
2. It just felt like the natural home for this check

I don't know if it technically counts as dead code elimination or if it's more appropriately const folding? Or perhaps it is commonly referred to as something else altogether? If so, I'll move it or break it out into it's own pass 😄

Reply:
So one concern here is that we are overlooking Boolean exprs that could be optimized somewhere else. A simple example would be:
```rust
var y: int;
state x = true && y;
```
Your current implementation wouldn't catch that. What I'm saying is that, similar to const folding, we need to look at all expressions here, not just constraints. And technically these "Boolean identities" should be handled in constant folding.

That being said, I think we should expand the scope of this a little bit and include stuff arithmetic identities. Example:
```rust
x * 0 == 0
0 / x == 0
x * 1 == x
x / 1 == x
x + 0 == 0
x - 0 == 0
x % 1 == x
```
(there may be others.. think through them)

Reply:
Just moved boolean identity folding to `const_folding`. Will add arithmetic identities next